pkg/osdir: add package for well-known directories

gcurtis opened the pull request:
Package osdir provides a cross-platform interface for well-known system and user directories.

The pre-defined DirType package variables use default paths that are specific to the current operating system. The standard XDG environment variables (XDG_CACHE_HOME, XDG_CONFIG_HOME, XDG_DATA_HOME, XDG_STATE_HOME) override these paths when set to a non-empty value, even on macOS and Windows.

Below are the definitions for the default directories on each system.

Linux:

```go
var Cache = DirType{
        System: "/var/cache",
        User:   "$XDG_CACHE_HOME", UserDefault: "$HOME/.cache",
        PrefixHint: "$CACHE_DIRECTORY",
}
var Config = DirType{
        System: "/etc",
        User:   "$XDG_CONFIG_HOME", UserDefault: "$HOME/.config",
        Search: "$XDG_CONFIG_DIRS", SearchDefault: "/etc/xdg",
        PrefixHint: "$CONFIGURATION_DIRECTORY",
}
var Data = DirType{
        System: "/usr/share",
        User:   "$XDG_DATA_HOME", UserDefault: "$HOME/.local/share",
        Search: "$XDG_DATA_DIRS", SearchDefault: "/usr/local/share:/usr/share",
}
var State = DirType{
        System: "/var/lib",
        User:   "$XDG_STATE_HOME", UserDefault: "$HOME/.local/state",
        PrefixHint: "$STATE_DIRECTORY",
}
```

Darwin:

```go
var Cache = DirType{
        System: "/Library/Caches",
        User:   "$XDG_CACHE_HOME", UserDefault: "$HOME/Library/Caches",
}
var Config = DirType{
        System: "/Library/Application Support",
        User:   "$XDG_CONFIG_HOME", UserDefault: "$HOME/Library/Application Support",
}
var Data = DirType{
        System: "/Library/Application Support",
        User:   "$XDG_DATA_HOME", UserDefault: "$HOME/Library/Application Support",
}
var State = DirType{
        System: "/Library/Application Support",
        User:   "$XDG_STATE_HOME", UserDefault: "$HOME/Library/Application Support",
}
```

Windows:

```go
var Cache = DirType{
        System: "$ProgramData",
        User:   "$XDG_CACHE_HOME", UserDefault: "$LocalAppData",
}
var Config = DirType{
        System: "$ProgramData",
        User:   "$XDG_CONFIG_HOME", UserDefault: "$AppData",
}
var Data = DirType{
        System: "$ProgramData",
        User:   "$XDG_DATA_HOME", UserDefault: "$AppData",
}
var State = DirType{
        System: "$ProgramData",
        User:   "$XDG_STATE_HOME", UserDefault: "$LocalAppData",
}
```

mikeland73 reviewed (commented):
makes me a bit nervous to implement so much of these ourselves (vs just relying on defaults).

Depending on what of this package we need, wouldn't something simpler like

```golang
func UserCacheDir() (string, error) {
	// First check XDG_CACHE_HOME
	if cacheDir := os.Getenv("XDG_CACHE_HOME"); cacheDir != "" {
		return cacheDir, nil
	}

	return os.UserCacheDir()
}
```

be easier to maintain and work for us? Or am I missing some more complex use cases we need?

loreto commented:
Do we need the other XDG defined dirs? https://specifications.freedesktop.org/basedir-spec/latest/
(DataDir, ConfigDir, etc)

Some of the other helper functions in here besides the ones that resolve the XDG dirs remind me of https://github.com/jetify-com/opensource/tree/main/pkg/fileutil and https://pkg.go.dev/path/filepath

(Just pointing it out ... but don't want to hold up things too much if these are needed; another path could be: move the truly common ones in here to OSS, and keep the ones you need for machines where they currently are? 🤷 )

gcurtis commented on pkg/osdir/osdir.go:
```diff
@@ -0,0 +1,341 @@
+// JoinPath joins path to the system or user directory of d and returns the
+// resulting absolute path. The given path must be relative and cannot contain
+// any ".." elements.
+func (d DirType) JoinPath(path string) (string, error) {
```
I'm not sure if we need it. The `xdg` package only exposes `*Subpath` functions because we never really need the path to the directory itself.

gcurtis commented:
> Do we need the other XDG defined dirs? https://specifications.freedesktop.org/basedir-spec/latest/
> (DataDir, ConfigDir, etc)

I think the only one we might not use today is `Data` (`XDG_DATA_HOME`). Are you asking to see if we could remove some?

> Some of the other helper functions in here besides the ones that resolve the XDG dirs remind me of https://github.com/jetify-com/opensource/tree/main/pkg/fileutil and https://pkg.go.dev/path/filepath

I can remove the ones like `MkdirAll`, `Remove`, etc. to pare things down. They're just conveniences that perform `JoinPath` + some file operation.

`DirType.ReadFile` does do some extra searching. For example, `Config.ReadFile` will try the paths in `XDG_CONFIG_DIRS` if a file can't be found in `XDG_CONFIG_HOME`. Some applications use this to allow a system-wide config that can be overridden with a user-specific config. You could argue we don't care about that though, in which case I'd just remove `XDG_CONFIG_DIRS`, `XDG_DATA_DIRS` and `JoinSearchPath`.

The `fileutil` package doesn't compile on Windows due to its dependencies, so that couldn't be used here.

gcurtis commented:
- Removed the `MkdirAll`, `ReadTextFile`, `RemoveFile`, and `WriteTextFile` methods.
- Removed the `PrefixHint` field.
- Exported `DirType.OpenFile`.

loreto reviewed (approved):
LGTM
